install-validator/pkg/dir/upgrade.go

```diff
@@ -0,0 +1,67 @@
+package dir
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+
+	kubewrapper "github.com/vmware-tanzu/graph-framework-for-microservices/install-validator/pkg/k8s-utils"
+)
+
+// ApplyDir checks if there are any incompatible crds and data for them. Applies them based on force flag.
+func ApplyDir(directory string, force bool, c kubewrapper.ClientInt, cFunc compareFunc) error {
+	err := c.FetchCrds()
+	if err != nil {
+		logrus.Error(err)
+	}
+
+	// check for incompatible models and not installed. Return  if any and force != true
+	inCompatibleCRDs, err := CheckDir(directory, c, cFunc)
+	if err != nil {
+		return err
+	}
+	if len(inCompatibleCRDs) > 0 && !force {
+		textChanges := new(bytes.Buffer)
+		for _, txt := range inCompatibleCRDs {
+			textChanges.Write(txt.Bytes())
+		}
+		logrus.Warn(textChanges)
+		return errors.New("incompatible datamodel changes detected")
+	}
+
+	// check if any data for incompatible models and return if so
```

Comment:
Can we replace data with crd ? it is confusing what data is?

Can we make the comment itself clear to anyone who sees the code.

Reply:
Its actually about CR in CRDs - we cannot upgrade if there are any even, if we use force. Updated comment to be more specific